### ISSUE TITLE
Runtime option to analyze single file

### ISSUE THREAD

**Issue by benjaminwood** [codeclimate/codeclimate#62]
**Background**: The CLI is awesome and I'm sure people are applying it in all sorts of scenarios. The way I see myself using it most often is while trying to resolve issues in a hot spot. 

**Scenerio**: Say I receive a notification that a file I just committed fell from an A to a C rating. It'd be useful if I could use the CLI locally on my machine to re-analyze that file while I'm working on it to see that I'm resolving the issues.

**Problem**: To my knowledge this isn't currently possible because the CLI doesn't offer the ability to analyze a single file.

**Potential solution**: Add a runtime option to the analyze command which allows the user to specify the relative path of the file they'd like to analyze.

Thoughts/ideas on how to implement this?

**Comment by GordonDiggs** [comment]
I agree this would be a good feature! The `-f` argument is already taken for the formatter, but I see two options:

* Use `-a` /  `--analyze` and you pass the filename there
* Accept it as the last argument to the command `codeclimate analyze -f json path/to/file`

I lean towards the second personally

**Comment by ktowle** [comment]
+1 on the ability to analyze a single file, or perhaps also a single directory

**Comment by brynary** [comment]
Quick update: There's been movement on this.

@ABaldwinHunter  -- Can you provide an update here?

**Comment by ABaldwinHunter** [comment]
@brynary @benjaminwood Progress is indeed underway! 

tl;dr By the end of this week, we should be able to release a new version of the CLI supporting this feature.

We have a [drafted solution](https://github.com/codeclimate/codeclimate/pull/113/files) that's close to being ready for release.

**Some details about implementation**: 

In the past, the CLI passed each engine present in `.codeclimate.yml`: a code directory, engine config, and list of `exclude_paths`. We've recently switched to an implementation (in part because of permissions issues on some files), that involves passing each engine `include_paths` to analyze (derived from the excludes) instead.

This change has made it much simpler to implement runtime analysis of particular files passed as args to `codeclimate analyze`. While some engines have already been updated to accept `include_paths` ([codeclimate-eslint](https://github.com/codeclimate/codeclimate-eslint/commit/0e39d21b4c8775ac0b60e0c9a37f8814a2f3f4f5) and [codeclimate-rubocop](https://github.com/codeclimate/codeclimate-rubocop/commit/9f1e178b0a826d04ade57d28e7cb2833cabc5558) for instance), a few others still need updating.

**Comment by ABaldwinHunter** [comment]
Update: this feature is now available with [release 0.5.0](https://github.com/codeclimate/codeclimate/releases).

Caveat: only engines that have been updated to support handling of `include_paths` will support this feature.

Engines that _do_ support analysis of single files include `codeclimate-rubocop` and `codeclimate-eslint`.

**Comment by ABaldwinHunter** [comment]
Going to close this issue but please feel free to comment or reach out!

**Comment by benjaminwood** [comment]
Sweet, that's awesome news! Thanks @ABaldwinHunter for the updates and do the rest of the team that worked on the feature.
